Fix: Improve error handling for Server API

kfahad5607 opened the pull request:
**Fixes #141 by checking for entities before deleting/modifying them to avoid prisma query error.**

#### Route handler changes
- Checking if team membership exists in **Remove a user from a team**
- Checking if team and user exist in **Grant a team permission to a user**
- Checking if team-permission for user exists in **Revoke a team permission from a user**
- Checking if team exists in **Delete a team**
- Checking if user exists in **Update user**
- Checking if team membership exists when `selected_team_id` is passed in request body in **Update user**
- Checking if user exists in **Delete user**

#### Other changes
- Updated `selectedTeamIdSchema` to validate string as `uuid`
- Added new helper function `ensureUserTeamPermissionExist` in `request-checks` to use in **Revoke a team permission from a user**
- Added new error `UserTeamPermissionNotFound` to use in `ensureUserTeamPermissionExist`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/stack-auth/stack?pullRequest=170) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/stack-auth/stack?pullRequest=170) before we can accept your contribution.<br/><hr/>**Fahad Khan** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/stack-auth/stack?pullRequest=170) it.</sub>

fomalhautb commented on apps/backend/src/lib/request-checks.tsx:
```diff
@@ -105,6 +105,67 @@ export async function ensureUserHasTeamPermission(
   }
 }
 
+export async function ensureUserTeamPermissionExist(
```
I think this function is a little bit duplicated to `ensureUserHasTeamPermission`

kfahad5607 replied:
I am combining these functions, but logically, they signify different types of errors.

1. `ensureUserHasTeamPermission`
   - This function throws a `TeamPermissionRequired` error, which indicates an authorization error. Users might misconstrue this error as a permission issue in the case of the **Revoke a team permission from a user** API request.

2. `ensureUserTeamPermissionExist`
   - This function throws a `UserTeamPermissionNotFound` error, indicating that the permission the user is trying to delete does not exist. This is more appropriate for the **Revoke a team permission from a user** API request.

fomalhautb replied:
yeah, I added `errorType` parameter to decide which type of error to throw in the `ensureUserHasTeamPermission` function.

fomalhautb commented on apps/backend/src/app/api/v1/users/crud.tsx:
```diff
@@ -158,6 +167,11 @@ export const usersCrudHandlers = createLazyProxy(() => createCrudHandlers(usersC
 
+      // If the selected_team_id is present and we reach here that means user does exist. Hence, we have this check.
+      if (!data.selected_team_id) {
+        await ensureUserExist(tx, { projectId: auth.project.id, userId: params.user_id });
```
We can move this check to the beginning and remove the condition so it is easier to follow

fomalhautb reviewed (changes requested):
Looks pretty clean!

kfahad5607 commented:
I have made the requested changes.

- Deleted the `ensureUserTeamPermissionExist` function and `UserTeamPermissionNotFound` error.
- Updated the `ensureUserHasTeamPermission` function to accept both **System Permission** and **Custom Permission** via `options.permissionId`.
- Checking for an existing membership between the team and user, instead of performing individual checks for the user and team in **Grant a team permission to a user**.

PS: I have also signed the CLA.

N2D4 commented:
> [![CLA assistant check](https://camo.githubusercontent.com/fbf2b4a56b8f8dacb3fb4f7f3cf7877b8c27f9cfee8546ff24c8df94facca0f6/68747470733a2f2f636c612d617373697374616e742e696f2f70756c6c2f62616467652f6e6f745f7369676e6564)](https://cla-assistant.io/stack-auth/stack?pullRequest=170)

FYI, looks like the bot hasn't picked up the CLA, but I can confirm that you signed it, so all good!

fomalhautb commented:
Thanks for the contribution!
